File: 6-22-剑指04-二维数组的查找/findNumberIn2DArray.py
#思路一：暴力解决，时间复杂度O(n)
#思路二：观察到二维数组从左至右、从上到下，数值都是递增的。所以先遍历右上角元素，若target比其小，
# 则往左查找；若target比其大，则往下查找

# class Solution:
#     def findNumberIn2DArray(self, matrix, target: int) -> bool:
#         for i in matrix:
#             for j in i:
#                 if target == j:
#                     return True
#         return False

class Solution:
    def findNumberIn2DArray(self, matrix, target: int) -> bool:

        if matrix == [] or len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        n = len(matrix)
        m = len(matrix[0])
        i = 0
        j = m - 1
        while i < n and j >= 0:
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] > target:
                j = j - 1
            else:
                i = i + 1
        return False
obj = Solution()
matrix = [

]
A = obj.findNumberIn2DArray(matrix, 20)
print(A)

File: Java/src/剑指OFFER/4-二维数组中的查找/FindNumberIn2DArray.java
public class FindNumberIn2DArray {

    public static void main(String[] args) {
        System.out.println("solution");

    }
    //这一题的着手点在于不能按照按照常规思路从左上方开始查；应该发现数字分布规律，从右上开始
    public static boolean findNumberIn2DArray(int[][] matrix, int target) {
        int len1 = matrix.length;
        int len2,i = 0,j = 0;
        if(len1 > 0){
            len2 = matrix[0].length;
            j = len2 - 1;
        }
        while (i<len1 && j>=0){
            if(matrix[i][j] == target)
                return true;
            else if(matrix[i][j] > target){ //第j列的所有数都大于target
                j--;
            }else {
                i++;
            }
        }
        return false;
    }
}
